### PR TITLE
refactor: internal handling of imports

### PR DESCRIPTION
**Pull request by charles-cooper**
### Commit message

```
this commit refactors how imports are handled internally.

historically, vyper handled imports by using a preprocessing step
(`extract_file_interface_imports`) which resolved imports to files and
provided them to the compiler pipeline as pure inputs. however, this
causes problems once the module system gets more complicated:
- it mixes passes. resolving imports and loading the files should
  essentially be resolved during analysis, but instead they are being
  resolved before the compiler is even entered into(!)
- it produces slightly different code paths into the main compiler entry
  point which introduces subtle bugs over time from scaffolding
  differences
- relatedly, each entry point into the compiler has to maintain its own
  mechanism for resolving different kinds of inputs to the compiler
  (JSON interfaces vs .vy files at the moment).

this commit replaces the external scaffolding with an "InputBundle"
abstraction which essentially models how the compiler interacts with its
inputs (depending on whether it is using the filesystem or JSON inputs).
this makes the entry points to the compiler overall simpler, and have
more consistent behavior.

this commit also:
- changes builtin interfaces so they are represented in the codebase as
  `.vy` files which are imported similarly to how regular (non-builtin)
  files are imported
- simplifies the `compile_files` and `compile_json` pipelines
- removes the `compile_codes` API, which was not actually more useful
  than the `compile_code` API (which is kept).
- cleans up tests by introducing a `make_file` and `make_input_bundle`
  abstraction
- simplifies and merges several files in the tests/cli/ directories
- adds a test for multiple output selections in the standard json pipeline
```

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/compiler/input_bundle.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+# stub
+@dataclass
+class VyFile(CompilerInput):
+    source_code: str
+
+
+# stub
+@dataclass
+class ABIInput(CompilerInput):
+    # some json file, either regular ABI or ethPM manifest v3 (EIP-2687)
```
Comment (fubuloubu) [inline vyper/compiler/input_bundle.py:22]:
```suggestion
    # some json file, either regular ABI or ethPM manifest v3 (EIP-2678)
```

File: vyper/compiler/input_bundle.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
+from dataclasses import dataclass
+from typing import Any
+
+from pathlib import Path, PurePath
+
+
+# stub
+class CompilerInput:
```
Comment (fubuloubu) [inline vyper/compiler/input_bundle.py:8]: Maybe `ImportedFile`

File: vyper/compiler/input_bundle.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+# stub
+@dataclass
+class ABIInput(CompilerInput):
+    # some json file, either regular ABI or ethPM manifest v3 (EIP-2687)
+    abi: Any
+
+
+@dataclass
+class InputBundle:
```
Comment (fubuloubu) [inline vyper/compiler/input_bundle.py:27]: Maybe `Imports` or `ImportedFiles`

File: vyper/compiler/input_bundle.py (hunk 2 above)
Comment (charles-cooper) [reply vyper/compiler/input_bundle.py:10]: it's any input file, including a compilation target

File: vyper/compiler/input_bundle.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    search_paths: list[Path]
+    # compilation_targets: dict[str, str]  # contract names => contract sources
+
+    def _load_from_path(self, path):
+        raise NotImplementedError(f"not implemented! {self.__class__}._load_from_path()")
+
+    def load_file(self, path: Path) -> str:
+        for p in self.search_paths:
```
[the remark this answers is not in the document]
Comment (charles-cooper) [reply vyper/compiler/input_bundle.py:43]: it has a return in the body.

**Review by fubuloubu — APPROVED** [review]

**Pull request merged**